src/tachimawari/control/sdk/packet/model/group_bulk_read.cpp

```diff
@@ -18,70 +18,69 @@
 // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 // THE SOFTWARE.
 
+#include "tachimawari/control/sdk/packet/model/group_bulk_read.hpp"
+
 #include <map>
 #include <memory>
 #include <string>
-#include <vector>
 #include <typeinfo>
-
-#include "tachimawari/control/sdk/packet/model/group_bulk_read.hpp"
-
-#include "tachimawari/joint/protocol_1/mx28_address.hpp"
-#include "tachimawari/control/sdk/module/marin_core_address.hpp"
+#include <vector>
 
 #include "dynamixel_sdk/dynamixel_sdk.h"
+#include "tachimawari/control/sdk/module/marin_core_address.hpp"
+#include "tachimawari/joint/protocol_1/mx28_address.hpp"
 
 namespace tachimawari::control::sdk
 {
 
 void GroupBulkRead::insert_all(
-  std::shared_ptr<std::map<uint8_t, std::shared_ptr<sdk::GroupBulkRead>>> bulk_data,
-  std::shared_ptr<sdk::GroupBulkRead> sdk_group_bulk_read)
+  std::shared_ptr<std::unordered_map<uint8_t, std::vector<uint8_t>>> bulk_data,
+  const std::shared_ptr<sdk::GroupBulkRead> & sdk_group_bulk_read)
 {
-  for (auto id : sdk_group_bulk_read->get_parameters_id()) {
-    if (bulk_data->find(id) != bulk_data->end()) {
-      bulk_data->insert({id, sdk_group_bulk_read});
-      (*bulk_data)[id] = sdk_group_bulk_read;
+  for (auto param : sdk_group_bulk_read->get_parameters()) {
+    std::vector<uint8_t> data(param.starting_address + param.length, 0);
+    auto param_data = sdk_group_bulk_read->get_from_param(param);
+
+    if (bulk_data->find(param.id) != bulk_data->end()) {
+      data = bulk_data->at(param.id);
+    }
+
+    for (size_t i = 0; i < param_data.size(); ++i) {
+      data[param.starting_address + i] = param_data[i];
+    }
+
+    if (bulk_data->find(param.id) != bulk_data->end()) {
+      (*bulk_data)[param.id] = data;
+    } else {
+      bulk_data->insert({param.id, data});
     }
```

Comment:
Do you guys support C++17? There's a short hand to insert or assign a map in C++ using [`insert_or_assign`](https://en.cppreference.com/w/cpp/container/unordered_map/insert_or_assign)

```suggestion
    bulk_data->insert_or_assign(param.id, data);
```